src/codegen/interpreter/bytecode_builder.cpp

```diff
@@ -0,0 +1,1830 @@
+}
+
+Opcode BytecodeBuilder::GetOpcodeForTypeSizeIntTypes(Opcode untyped_op,
+                                                     llvm::Type *type) const {
+  index_t id = BytecodeFunction::GetOpcodeId(untyped_op);
+
+  // This function highly depends on the macros in bytecode_instructions.def!
+
+  switch (code_context_.GetTypeSize(type)) {
+    case 1:
+      return BytecodeFunction::GetOpcodeFromId(id + 0);
+    case 2:
+      return BytecodeFunction::GetOpcodeFromId(id + 1);
+    case 4:
+      return BytecodeFunction::GetOpcodeFromId(id + 2);
+    case 8:
+      return BytecodeFunction::GetOpcodeFromId(id + 3);
+    default:
+      throw NotSupportedException("llvm type size not supported: " +
+                                  CodeGen::Dump(type));
+  }
+}
+
+Instruction &BytecodeBuilder::InsertBytecodeInstruction(
+    const llvm::Instruction *llvm_instruction, Opcode opcode,
+    const std::vector<index_t> &args) {
+  PELOTON_ASSERT(opcode != Opcode::undefined);
+
+  // calculate number of required instruction slots
+  const size_t number_instruction_slots =
+      ((2 * (1 + args.size())) + sizeof(instr_slot_t) - 1) /
+      sizeof(instr_slot_t);
+
+  bytecode_function_.bytecode_.insert(bytecode_function_.bytecode_.end(),
+                                      number_instruction_slots, 0);
+  Instruction &instruction = *reinterpret_cast<Instruction *>(
+      &*(bytecode_function_.bytecode_.end() - number_instruction_slots));
+  instruction.op = opcode;
+  for (size_t i = 0; i < args.size(); i++) instruction.args[i] = args[i];
+
+  AddInstructionToTrace(llvm_instruction, number_instruction_slots);
+
+  return instruction;
+}
+
+Instruction &BytecodeBuilder::InsertBytecodeInstruction(
+    const llvm::Instruction *llvm_instruction, Opcode opcode,
+    const std::vector<const llvm::Value *> &args) {
+  PELOTON_ASSERT(opcode != Opcode::undefined);
+
+  std::vector<index_t> args_transformed(args.size());
+  std::transform(
+      args.begin(), args.end(), args_transformed.begin(),
+      [this](const llvm::Value *value) { return GetValueSlot(value); });
+
+  return InsertBytecodeInstruction(llvm_instruction, opcode, args_transformed);
+}
+
+ExternalCallInstruction &BytecodeBuilder::InsertBytecodeExternalCallInstruction(
+    const llvm::Instruction *llvm_instruction, index_t call_context,
+    void *function) {
+  // calculate number of slots and assert it is 2 (this way we recognise if any
+  // size changes)
+  const size_t number_instruction_slots =
+      (sizeof(ExternalCallInstruction) + sizeof(instr_slot_t) - 1) /
+      sizeof(instr_slot_t);
+  PELOTON_ASSERT(number_instruction_slots == 2);
+
+  bytecode_function_.bytecode_.insert(bytecode_function_.bytecode_.end(),
+                                      number_instruction_slots, 0);
+
+  ExternalCallInstruction instruction = {
+      Opcode::call_external, call_context,
+      reinterpret_cast<void (*)(void)>(function)};
+  *reinterpret_cast<ExternalCallInstruction *>(
```

Comment:
This is a bit messy, can you pull out the pointer and perform the assignment on two separate lines?